tests/nvmeof/test_ceph_nvmeof_gateway_sub_scale.py

```diff
@@ -0,0 +1,265 @@
+
+from ceph.ceph import Ceph
+from ceph.ceph_admin import CephAdmin
+from ceph.ceph_admin.common import fetch_method
+from ceph.nvmeof.initiator import Initiator
+from ceph.nvmeof.nvmeof_gwcli import NVMeCLI, find_client_daemon_id
+from ceph.utils import get_node_by_id
+from tests.rbd.rbd_utils import initial_rbd_config
+from utility.log import Log
+from utility.retry import retry
+from utility.utils import generate_unique_id, run_fio
+
+LOG = Log(__name__)
+
+
+def initiators(ceph_cluster, gateway, config):
+    """Configure NVMe Initiators."""
+    client = get_node_by_id(ceph_cluster, config["node"])
+    initiator = Initiator(client)
+    initiator.disconnect_all()
+
+    connect_cmd_args = {
+        "transport": "tcp",
+        "traddr": gateway.ip_address,
+        "trsvcid": config["listener_port"],
+        "nqn": config["subnqn"],
+    }
+    LOG.debug(initiator.connect(**connect_cmd_args))
+
+
+@retry(Exception, tries=5, delay=10)
+def run_io(ceph_cluster, num, io):
+    """Run IO on newly added namespace."""
+    LOG.info(io)
+    client = get_node_by_id(ceph_cluster, io["node"])
+    initiator = Initiator(client)
+
+    targets = initiator.list_spdk_drives()
+    if not targets:
+        raise Exception(f"NVMe Targets not found on {client.hostname}")
+
+    device_path = next(
+        (device["DevicePath"] for device in targets if device["NameSpace"] == num), None
+    )
+    LOG.debug(device_path)
+
+    if device_path is None:
+        raise Exception("NVMe volume not found")
+    else:
+        io_args = {
+            "device_name": device_path,
+            "client_node": client,
+            "run_time": "10",
+            "long_running": True,
+            "io_type": io["io_type"],
+            "cmd_timeout": "notimeout",
+        }
+        result = run_fio(**io_args)
+        if result != 0:
+            raise Exception("FIO failure")
+
+
+def configure_subsystem(config, nvme_cli, ceph_cluster, node):
+    max_ns = config["args"].pop("max-namespaces")
+    pool_name = config["args"].pop("pool")
+    port = config["args"].pop("port")
+    hostnqn = config["args"].pop("hostnqn")
+
+    for num in range(1, config["args"].pop("subsystems") + 1):
+        config["args"].update(
+            {
+                "subnqn": f"nqn.2016-06.io.spdk:cnode{num}",
+                "serial_num": f"{num}",
+                "m": max_ns,
+            }
+        )
+
+        subsystem_func = fetch_method(nvme_cli, "create_subsystem")
+        subsystem_func(**config["args"])
+        config["args"].clear()
+
+        client_id = find_client_daemon_id(
+            ceph_cluster, pool_name, node_name=node.hostname
+        )
+        config["args"].update(
+            {
+                "subnqn": f"nqn.2016-06.io.spdk:cnode{num}",
+                "gateway-name": client_id,
+                "traddr": node.ip_address,
+                "port": port,
+            }
+        )
+
+        listener_func = fetch_method(nvme_cli, "create_listener")
```

Comment:
Its better to have these functionalities in a seperate blocks and call them from L239. 
I am approving PR now, please keep this point in mind and implement it whenever you have some bandwidth.

Reply:
@HaruChebrolu I gave a thought about this for another time but I think this implementation looks good as it is `configure_subsystem` and not `create_subsystem`. So, if say any test calls to configure a subsystem entirely, this snippet will do E2E. If test requires only to add listener then it points to https://github.com/red-hat-storage/cephci/pull/3301/files#diff-aef30ea33d621d41fe3858267636fde0c522e9d3aedf3205fcdc8588d8cdf048R111

Reply:
okay